pkg/util/asyncworker/async_limited_workers.go

```diff
@@ -0,0 +1,245 @@
+		ctx = context.WithValue(ctx, contextKeyMetricEmitter, alw.emitter)
+	}
+
+	status := &workStatus{
+		working:   true,
+		startedAt: time.Now(),
+		work:      work,
+	}
+	status.ctx, status.cancelFn = context.WithCancel(ctx)
+	alw.activeQ[work.Name] = status
+	return status.ctx
+}
+
+func (alw *AsyncLimitedWorkers) doHandle(ctx context.Context, work *Work) {
+	var handleErr error
+
+	defer func() {
+		if r := recover(); r != nil {
+			handleErr = fmt.Errorf("recover from %v", r)
+
+			metricErr := EmitCustomizedAsyncedMetrics(ctx,
+				metricNameAsyncWorkPanic, 1,
+				metrics.ConvertMapToTags(map[string]string{
+					"workName": work.Name,
+				})...)
+
+			if metricErr != nil {
+				general.Errorf("emit metric(%s:%d) failed with err: %v",
+					metricNameAsyncWorkDurationMs, 1, metricErr)
+			}
+		}
+
+		alw.complete(work, handleErr)
+	}()
+
+	general.InfoS("handle work",
+		"AsyncLimitedWorkers", alw.name,
+		"workName", work.Name,
+		"params", work.Params,
+		"deliveredAt", work.DeliveredAt)
+
+	funcValue := reflect.ValueOf(work.Fn)
+
+	// filling up parameters for the passed functions
+	paramValues := make([]reflect.Value, 1, len(work.Params)+1)
+	paramValues[0] = reflect.ValueOf(ctx)
+	for _, param := range work.Params {
+		paramValues = append(paramValues, reflect.ValueOf(param))
+	}
+
+	startTime := time.Now()
+	funcRets := funcValue.Call(paramValues)
+	workDurationMs := time.Since(startTime).Milliseconds()
+
+	if len(funcRets) != 1 {
+		handleErr = fmt.Errorf("work Fn returns invalid number: %d of return values", len(funcRets))
+	} else if funcRets[0].Interface() != nil {
+		var ok bool
+		handleErr, ok = funcRets[0].Interface().(error)
+
+		if !ok {
+			handleErr = fmt.Errorf("work Fn returns return value: %v of invalid type", funcRets[0].Interface())
+		}
+	}
+
+	metricErr := EmitCustomizedAsyncedMetrics(ctx,
+		metricNameAsyncWorkDurationMs, workDurationMs,
+		metrics.ConvertMapToTags(map[string]string{
+			"workName": work.Name,
+		})...)
+
+	if metricErr != nil {
+		general.Errorf("emit metric(%s:%d) failed with err: %v",
+			metricNameAsyncWorkDurationMs, workDurationMs, metricErr)
+	}
+}
+
+func (alw *AsyncLimitedWorkers) complete(completedWork *Work, workErr error) {
+	status := alw.activeQ[completedWork.Name]
+	status.finishedAt = time.Now()
+
+	general.InfoS("complete work",
+		"AsyncLimitedWorkers", alw.name,
+		"workName", completedWork.Name,
+		"params", completedWork.Params,
+		"deliveredAt", completedWork.DeliveredAt,
+		"startAt", status.startedAt,
+		"finishAt", status.finishedAt,
+		"workErr", workErr)
+
+	alw.workLock.Lock()
+	defer alw.workLock.Unlock()
+
+	work, ok := alw.backoffQ[completedWork.Name]
+	if ok {
+		alw.waitQ.Add(work)
+		delete(alw.backoffQ, completedWork.Name)
+	}
+	delete(alw.activeQ, completedWork.Name)
+
+	alw.cond.Broadcast()
+}
+
+func (alw *AsyncLimitedWorkers) Start(stopCh <-chan struct{}) error {
+	go wait.Until(alw.handle, 10*time.Second, stopCh)
```

Comment:
should we 
1. add another goroutine to select if stopCh closed, if so, triggering alw.cond.Broadcast()
2. perceive stopCh closed in doPoll -> poll -> handle, then stopping loop in handle